include/oneapi/dpl/experimental/kt/single_pass_scan.h

```diff
@@ -0,0 +1,413 @@
+          __atomic_partial_value(*(__partial_vals + __tile_id + __padding)),
+          __atomic_full_value(*(__full_vals + __tile_id + __padding))
+    {
+    }
+
+    void
+    set_partial(const _T __val)
+    {
+        __atomic_partial_value.store(__val, sycl::memory_order::release);
+        __atomic_flag.store(__partial_status, sycl::memory_order::release);
+    }
+
+    void
+    set_full(const _T __val)
+    {
+        __atomic_full_value.store(__val, sycl::memory_order::release);
+        __atomic_flag.store(__full_status, sycl::memory_order::release);
+    }
+
+    template <typename _Subgroup, typename _BinaryOp>
+    _T
+    cooperative_lookback(const _Subgroup& __subgroup, _BinaryOp __binary_op)
+    {
+        _T __running = oneapi::dpl::unseq_backend::__known_identity<_BinaryOp, _T>;
+        auto __local_id = __subgroup.get_local_id();
+
+        for (int __tile = static_cast<int>(__tile_id) - 1; __tile >= 0; __tile -= SUBGROUP_SIZE)
+        {
+            _AtomicFlagT __tile_flag_atomic(*(__flags_begin + __tile + __padding - __local_id));
+            _T __tile_flag = __initialized_status;
+
+            // Load flag from a previous tile based on my local id.
+            // Spin until every work-item in this subgroup reads a valid status
+            do
+            {
+                __tile_flag = __tile_flag_atomic.load(sycl::memory_order::acquire);
+            } while (!sycl::all_of_group(__subgroup, __tile_flag != __initialized_status));
+
+            bool __is_full = __tile_flag == __full_status;
+            auto __is_full_ballot = sycl::ext::oneapi::group_ballot(__subgroup, __is_full);
+            ::std::uint32_t __is_full_ballot_bits{};
+            __is_full_ballot.extract_bits(__is_full_ballot_bits);
+
+            _AtomicValueT __tile_value_atomic(
+                *((__is_full ? __full_vals_begin : __partial_vals_begin) + __tile + __padding - __local_id));
+            _T __tile_val = __tile_value_atomic.load(sycl::memory_order::acquire);
+
+            auto __lowest_item_with_full = sycl::ctz(__is_full_ballot_bits);
+            _T __contribution = __local_id <= __lowest_item_with_full
+                                    ? __tile_val
+                                    : oneapi::dpl::unseq_backend::__known_identity<_BinaryOp, _T>;
+
+            // Running reduction of all of the partial results from the tiles found, as well as the full contribution from the closest tile (if any)
+            __running = __binary_op(__running, sycl::reduce_over_group(__subgroup, __contribution, __binary_op));
+
+            // If we found a full value, we can stop looking at previous tiles. Otherwise,
+            // keep going through tiles until we either find a full tile or we've completely
+            // recomputed the prefix using partial values
+            if (__is_full_ballot_bits)
+                break;
+        }
+        return __running;
+    }
+
+    const uint32_t __tile_id;
+    _FlagType* __flags_begin;
```

Comment:
Since we have separated this from the partial / full scan values, should we use a smaller type than `uint32_t` as `_FlagType`?

Reply:
I believe that we are limited to 32bit /64 bit types with [atomics](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:atomic-references).  Also, not all devices have support for 64 bit atomics (FPGAs).

Reply:
Ah, I see. My comment can be ignored then.